src/EventSubscriber/InitSubscriber.php

```diff
@@ -0,0 +1,145 @@
+   *
+   * @throws \Drupal\Component\Plugin\Exception\InvalidPluginDefinitionException
+   * @throws \Drupal\Component\Plugin\Exception\PluginNotFoundException
+   */
+  public function __construct(AccountProxyInterface $account, DateFormatterInterface $date_formatter, EntityTypeManagerInterface $entity_type_manager, MessengerInterface $messenger, PasswordChecker $password_checker) {
+    $this->account = $account;
+    $this->dateFormatter = $date_formatter;
+    $this->messenger = $messenger;
+    $this->passwordChecker = $password_checker;
+    $this->policyStorage = $entity_type_manager->getStorage('password_enhancements_policy');
+    $this->userStorage = $entity_type_manager->getStorage('user');
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function getSubscribedEvents(): array {
+    $events[KernelEvents::REQUEST][] = ['checkRequiredPasswordChange', 32];
+    $events[KernelEvents::REQUEST][] = ['passwordChangeNotification', 34];
+    return $events;
+  }
+
+  /**
+   * Checks if the user has to change the password already or not.
+   *
+   * Sets the password change required field to true to force password change if
+   * the user's password expired.
+   *
+   * @param \Symfony\Component\HttpKernel\Event\GetResponseEvent $event
+   *   Response event.
+   *
+   * @throws \Drupal\Core\Entity\EntityStorageException
+   *   If the user was failed to save.
+   */
+  public function checkRequiredPasswordChange(GetResponseEvent $event) {
+    if ($event->isMasterRequest()) {
+      $policy = $this->policyStorage->loadByRoleAndPriority($this->account->getRoles());
+      if ($policy !== NULL) {
+        $user = $this->userStorage->load($this->account->id());
+        if (!$user->password_enhancements_password_change_required->value && $this->passwordChecker->isExpired($policy)) {
+          $user->password_enhancements_password_change_required->value = TRUE;
+          $user->save();
+        }
+      }
+    }
+  }
+
+  /**
+   * Checks if the user's password is about to expire.
+   *
+   * If the user's password is about to expire it shows an error message if it
+   * is enabled.
+   *
+   * @param \Symfony\Component\HttpKernel\Event\GetResponseEvent $event
+   *   Response event.
+   */
+  public function passwordChangeNotification(GetResponseEvent $event) {
+    if ($event->isMasterRequest() && Url::fromRoute('password_enhancements.password_change')->toString() !== $event->getRequest()->getPathInfo()) {
+      $policy = $this->policyStorage->loadByRoleAndPriority($this->account->getRoles());
+      if ($policy !== NULL && $this->passwordChecker->showWarningMessage($policy)) {
+        $user = $this->userStorage->load($this->account->id());
+        $this->messenger->addWarning(new FormattableMarkup($policy->getExpiryWarningMessage(), [
+          '@date_time' => $this->dateFormatter->format($user->password_enhancements_password_changed_date->value + $policy->getExpireSeconds(), 'password_enhancements_date_format'),
```

Comment:
The purpose of this would be able to make it custom, that's why I didn't wanted to use any built-in formats or define a fixed custom format.

But why it is an issue? It's there, no one has to touch it, but if someone would want to change the format it is possible to do so, I would say it is a nice to allow it, also because the whole message is customizable too, so why not allow to customize date format itself?

Reply:
Because the date format should be consistent on a site and thanks for the pre-built date formats every core/contrib modules are using one of those. Also, site-builders are using those when they are creating a Views view for example.

This is the reason why I think we should not introduce this functionality right now, only when there is a demand for it. If we introduce it now and it stays until we release the stable - even if it is used by 1 user - we can not get rid of it easily.

Reply:
I would still keep it.